File: src/ios/OSLoggerProxyPlugin.m
#import "OSLoggerProxyPlugin.h"
#import "OSLogger.h"

@interface OSLoggerProxyPlugin ()

@property (nonatomic, readonly, strong) id <OSLoggerProtocol> logger;

@end

static OSDeviceInfo* deviceInfo;
@implementation OSLoggerProxyPlugin

#pragma mark - Life Cycle

-(void)pluginInitialize { 
    _logger = [OSLogger sharedInstance];
}

# pragma mark - Interface

+ (id)sharedInstance {
    static OSLoggerProxyPlugin *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (void)logMessageGeneral:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 2){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    
    [_logger logInfo:message withModule:moduleName];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

- (void)logMessageTrace:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 2){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    
    [_logger logVerbose:message withModule:moduleName];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

- (void)logMessageDebug:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 2){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    
    [_logger logDebug:message withModule:moduleName];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

- (void)logMessageWarning:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 2){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    
    [_logger logWarning:message withModule:moduleName];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

- (void)logMessageError:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 3){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    NSException *stack = [[NSException alloc] initWithName:@"MyException" reason:[[command arguments] objectAtIndex:2] userInfo:nil];
    
    [_logger logError:message withModule:moduleName withStack:stack];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

- (void)logMessageFatal:(CDVInvokedUrlCommand *)command{
    if([[command arguments] count] < 3){
        return;
    }
    
    NSString *message = [[command arguments] objectAtIndex:0];
    NSString *moduleName = [[command arguments] objectAtIndex:1];
    NSException *stack = [[NSException alloc] initWithName:@"MyException" reason:[[command arguments] objectAtIndex:2] userInfo:nil];
    
    [_logger logFatal:message withModule:moduleName withStack:stack];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
}

@end

File: src/android/OSLoggerProxyPlugin.java
package com.outsystems.plugins.loggerproxy;

import com.outsystems.plugins.oslogger.OSLogger;
import com.outsystems.plugins.oslogger.interfaces.Logger;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;

import org.json.JSONArray;
import org.json.JSONException;

public class OSLoggerProxyPlugin extends CordovaPlugin {

    private static final String ACTION_LOG_GENERAL = "logMessageGeneral";
    private static final String ACTION_LOG_TRACE = "logMessageTrace";
    private static final String ACTION_LOG_DEBUG = "logMessageDebug";
    private static final String ACTION_LOG_WARNING = "logMessageWarning";
    private static final String ACTION_LOG_ERROR = "logMessageError";
    private static final String ACTION_LOG_FATAL = "logMessageFatal";
    private static final String ACTION_NATIVE_CRASH = "forceNativeCrash";

    private Logger logger = OSLogger.getInstance();

    @Override
    protected void pluginInitialize() { }

    @Override
    public boolean execute(String action, JSONArray args, final CallbackContext callbackContext) throws JSONException {
        if (ACTION_LOG_GENERAL.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            this.logger.logInfo(message, moduleName);

            return true;
        }
        else if (ACTION_LOG_TRACE.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            this.logger.logVerbose(message, moduleName);

            return true;
        }
        else if (ACTION_LOG_DEBUG.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            this.logger.logDebug(message, moduleName);

            return true;
        }
        else if (ACTION_LOG_WARNING.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            this.logger.logWarning(message, moduleName);
            
            return true;
        }
        else if (ACTION_LOG_ERROR.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            Exception e = new Exception(args.getString(2));
            this.logger.logError(message, moduleName, e);

            return true;
        }
        else if (ACTION_LOG_FATAL.equals(action)) {
            String message = args.getString(0);
            String moduleName = args.getString(1);
            Exception e = new Exception(args.getString(2));
            this.logger.logFatal(message, moduleName, e);

            return true;
        }
        else if(ACTION_NATIVE_CRASH.equals(action)){
            throw new RuntimeException("This is a crash");
        }    

        return false;
    }
}